modules/diskq/logqueue-disk-non-reliable.c

```diff
@@ -649,6 +651,139 @@ _push_tail(LogQueue *s, LogMessage *msg, const LogPathOptions *path_options)
+          self->input_queues[thread_index].finish_cb_registered = TRUE;
+          log_queue_ref(&self->super.super);
+        }
+
+      log_msg_write_protect(msg);
+      node = log_msg_alloc_queue_node(msg, path_options);
+      iv_list_add_tail(&node->list, &self->input_queues[thread_index].items);
+
+      self->input_queues[thread_index].len++;
+
+      log_msg_unref(msg);
+      return;
+    }
+  // slow path
+  _push_tail_single_message(s, msg, path_options);
+}
+
+static void
+_move_part_of_input_to_front_cache(LogQueueDiskNonReliable *self, InputQueue *input_queue,
+                                   gint num_msgs_to_send_to_front_cache)
+{
+  struct iv_list_head *head = &input_queue->items;
+  struct iv_list_head *pos = head->next;
+  gint count = 0;
+  g_mutex_lock(&self->super.super.lock);
+  while (pos != head && count < num_msgs_to_send_to_front_cache)
+    {
+      struct iv_list_head *next = pos->next;
+      LogMessageQueueNode *node = iv_list_entry(pos, LogMessageQueueNode, list);
+      LogPathOptions lpo = LOG_PATH_OPTIONS_INIT;
+      lpo.ack_needed = node->ack_needed;
+      lpo.flow_control_requested = node->flow_control_requested;
+
+      LogMessage *msg = log_msg_ref(node->msg);
+      _push_tail_front_cache(self, msg, &lpo);
+      log_msg_free_queue_node(node);
+
+      iv_list_del_init(pos);
+      input_queue->len--;
+      log_queue_queued_messages_inc(&self->super.super);
+
+      pos = next;
+      count++;
+    }
+  log_queue_push_notify(&self->super.super);
+  g_mutex_unlock(&self->super.super.lock);
+}
+
+static void
+_move_part_of_input_to_disk_or_flow_control_window(LogQueueDiskNonReliable *self, InputQueue *input_queue)
+{
+  LogMessageQueueNode *node = iv_list_entry(input_queue->items.next, LogMessageQueueNode, list);
+  LogPathOptions lpo = LOG_PATH_OPTIONS_INIT;
+  lpo.ack_needed = node->ack_needed;
+  lpo.flow_control_requested = node->flow_control_requested;
+  LogMessage *msg = log_msg_ref(node->msg);
+
+  _push_tail_single_message(&self->super.super, msg, &lpo);
+
+  iv_list_del_init(&node->list);
+  input_queue->len--;
+  log_msg_free_queue_node(node);
+}
+
+static gpointer
+_move_input(gpointer user_data)
+{
+  LogQueueDiskNonReliable *self = (LogQueueDiskNonReliable *) user_data;
+  gint thread_index = main_loop_worker_get_thread_index();
+  g_assert(thread_index >= 0);
+
+  if (self->input_queues[thread_index].len == 0)
+    goto exit;
+
+  while (self->input_queues[thread_index].len > 0)
+    {
+      gint num_msgs_to_send_to_front_cache = (qdisk_get_length(self->super.qdisk) == 0 && self->flow_control_window.len == 0)?
+                                             MIN(self->front_cache.limit - self->front_cache.len, self->input_queues[thread_index].len) : 0;
+
+      if (num_msgs_to_send_to_front_cache > 0)
+        {
+          _move_part_of_input_to_front_cache(self, &self->input_queues[thread_index], num_msgs_to_send_to_front_cache);
```

Comment:
do we want to do this multiple times? with the loop in place, we might be feeding the front cache  if the output is quick enough to get empty the front cache as we release the lock.

The answer is probably true, I was just wondering if this was intended.

Reply:
also, now as I was reading this a bit more, the else branch only moves a single message, so the condition will be evaluated for every message.

the condition seems to be racy, `front_cache.len` seems to be modified by the output thread without locking. Locking it would kill the entire purpose, and this race could be fine, but in this case, it should be spelled out why that is the case in a comment.